app/com/linkedin/drelephant/ElephantRunner.java

```diff
@@ -45,136 +48,377 @@
+    } catch (Exception e) {
+      logger.error("Error fetching job list. Try again later...", e);
+      return;
+    }
+
+    // There is a lag of job data from AM/NM to JobHistoryServer HDFS, we shouldn't use the current time, since
+    // there might be new jobs arriving after we fetch jobs. We provide one minute delay to address this lag.
+    for (int i = 0; i < completedJobs.size(); i++) {
+      completedJobs.get(i).updateExpiryTime(_fetchLag);
+    }
+
+    logger.info("Completed Job queue size is " + (_completedJobQueue.size() + completedJobs.size()));
+    logger.info("Undefined Job queue size is " + (_undefinedJobQueue.size() + undefinedJobs.size()));
 
+    _completedJobQueue.addAll(completedJobs);
+    _undefinedJobQueue.addAll(undefinedJobs);
+
+    _lastTime = to;
+  }
+
+  private class CompletedExecutorThread implements Runnable {
     private int _threadId;
-    private BlockingQueue<AnalyticJob> _jobQueue;
+    private DelayQueue<AnalyticJob> _completedJobQueue;
 
-    ExecutorThread(int threadNum, BlockingQueue<AnalyticJob> jobQueue) {
+    CompletedExecutorThread(int threadNum, DelayQueue<AnalyticJob> completedJobQueue) {
       this._threadId = threadNum;
-      this._jobQueue = jobQueue;
+      this._completedJobQueue = completedJobQueue;
     }
 
     @Override
     public void run() {
       while (_running.get() && !Thread.currentThread().isInterrupted()) {
         AnalyticJob analyticJob = null;
         try {
-          analyticJob = _jobQueue.take();
-          logger.info("Executor thread " + _threadId + " analyzing " + analyticJob.getAppType().getName() + " "
-              + analyticJob.getAppId());
-          AppResult result = analyticJob.getAnalysis();
-          result.save();
+          analyticJob =_completedJobQueue.take();
+          Job job = getJobFromCluster(analyticJob);
+          analyticJob.setJobStatus(job.getJobState().name()).setSeverity(0);
+          JobStatus.State jobState = job.getJobState();
 
+          // Job State can be RUNNING, SUCCEEDED, FAILED, PREP or KILLED
+          if (job.getJobState() == JobStatus.State.SUCCEEDED || job.getJobState() == JobStatus.State.FAILED) {
+            logger.info("Executor thread " + _threadId + " for completed jobs is analyzing "
+                + analyticJob.getAppType().getName() + " " + analyticJob.getAppId() + " :: " + jobState.name());
+            AppResult result = analyticJob.getAnalysis();
+            AppResult jobInDb = AppResult.find.byId(analyticJob.getAppId());
+            if (jobInDb == null) {
+              result.save();
+            } else {
+              // Delete and save to prevent Optimistic Locking
+              jobInDb.delete();
+              result.save();
+            }
+          } else {
+            // Should not reach here. We consider only SUCCEEDED and FAILED JOBs in _completedJobQueue
+            throw new RuntimeException(analyticJob.getAppId() + " is in " + jobState.name() + " state. This should not"
```

Comment:
Here we are analyzing jobs from the completedQueue. The completedJobs Queue will never have a non-completed job. So the code should never reach this section of code. However, if it does, then we need to fix the issue. But when it happens, we just want to retry it 3 times and then drop it.